### PR TITLE
Recipes - RecipeChoice support

### PR DESCRIPTION
**Pull request by Fusezion**
Start of recipe updates, RecipeChoice additions.

This PR adds support for ExactChoice, while updating some minor changes in syntax for readability and/or later use.

- Added error message when using SmithingRecipes on 1.20
  - Using the normal `new SmithingRecipe()` no longer works as intended and would just silently fail
  - Testing to see if this is paper only is needed, however you can't even place items in gui anyways
- Reworked MaterialChoice to use RecipeChoice to support ExactChoice
  - Materialchoice now properly supports `minecrafttag` classinfo causing a breaking change due to removal of old pattern
- Added expression to get choices of recipe choice, currently is limited in use until `ExprIngredientsOfRecipe` is changed to use RecipeChoice in a later pr. Additionally will need to decide if adding `random choice out of %recipechoice%` will be good in some format.
- Simplify effects to use new `getRecipeChoice/getExactChoice/getMaterialChoice` methods in `RecipeUtil` class removing unnecessary clutter 
- Updated Crafting and Cooking recipes to be split into seperate patterns for increased readability
  
image

### CODE REVIEW

File: src/main/java/com/shanebeestudios/skbee/api/recipe/RecipeUtil.java [hunk 1]
```diff
@@ -148,6 +157,146 @@ public static void logShapedRecipe(ShapedRecipe recipe) {
...
+        for (Object object : objects) {
+            if (object instanceof ExactChoice exactChoice) {
+                for (ItemStack itemStack : exactChoice.getChoices()) {
+                    if (itemStacks.contains(itemStack)) continue;
+                    itemStacks.add(itemStack);
+                }
+            } else if (object instanceof ItemStack itemStack) {
+                if (itemStacks.contains(itemStack)) continue;
+                boolean isAir = itemStack.getType().isAir();
+                if (isAir) continue;
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/api/recipe/RecipeUtil.java:211]: could this just be one line?
```suggestion
                if (itemStack.getType().isAir()) continue;
```
same with the one bellow

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
just a few requested changes

File: src/main/java/com/shanebeestudios/skbee/api/recipe/RecipeUtil.java (hunk 1 above)
Comment (Fusezion) [reply src/main/java/com/shanebeestudios/skbee/api/recipe/RecipeUtil.java:211]: I think this habit came from when I first redid these where the old code also split them all off into their own fields

File: src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCookingRecipe.java [hunk 2]
```diff
@@ -65,50 +68,34 @@ public class EffCookingRecipe extends Effect {
...
-        recipeType = parseResult.mark;
+    public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+        this.result = (Expression<ItemStack>) exprs[0];
+        this.ingredient = (Expression<Object>) exprs[1];
+        this.key = (Expression<String>) exprs[2];
+        this.experience = (Expression<Number>) exprs[3];
+        this.cookTime = (Expression<Timespan>) exprs[4];
+        this.group = (Expression<String>) exprs[5];
+        this.recipeType = matchedPattern;
+        Bukkit.broadcastMessage("" + matchedPattern);
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCookingRecipe.java:79]: Im assuming this was a debug?

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
ok, this is all I could find

File: src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCookingRecipe.java (hunk 2 above)
Comment (Fusezion) [reply src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCookingRecipe.java:79]: Ah yes I'm at work so I'll fix it when I get off in 4 hours unless you want to

Comment (ShaneBeee) [reply src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCookingRecipe.java:79]: yeah I can do it

File: src/main/java/com/shanebeestudios/skbee/api/recipe/RecipeUtil.java [hunk 3]
```diff
@@ -148,6 +157,143 @@ public static void logShapedRecipe(ShapedRecipe recipe) {
...
 
+    /**
+     * Get a recipe choice based off given argument
+     *
+     * @param object object to get RecipeChoice from
+     * @return RecipeChoice or null if invalid
+     */
+    @Nullable
+    public static RecipeChoice getRecipeChoice(Object object) {
+        if (object == null) return null;
```
Comment (Fusezion) [inline src/main/java/com/shanebeestudios/skbee/api/recipe/RecipeUtil.java:168]: Self reminder to update this to support `ItemType#getAll()`

**Comment by Fusezion** [comment]
Closing this pr and deleting branch to swap to my fork and account for any new changes. Don't feel like handling conflicts either.

**Pull request closed without merging**